### PR TITLE
Fix Dynamo tracing of compile support checks

### LINKED ISSUE #154259: `vision_maskrcnn` graph breaks increased from 18 to 20
As reported in #153985, vision_maskrcnn graph breaks is increased which blocked TorchVision pin upgrade. #154255 bumped the expected graph breaks to unblock. But we still need investigation on the increased #graph breaks.

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #185528

Newer torchvision roi_align calls torch._dynamo.utils.is_compile_supported(input.device.type) to choose between the Python ROIAlign path and the custom op path. Since torch._dynamo is skiplisted, Dynamo treated this helper as skipped and inserted graph breaks when vision_maskrcnn reached roi_align. That contributed to the torchvision pin graph-break regression.

Teach Dynamo's trace rules to represent is_compile_supported as a user function and mark the helper as a constant result. The helper depends on a constant device type and runtime compiler support, so Dynamo can evaluate it during tracing and guard the result instead of tracing into skiplisted internals. This is targeted to the one helper rather than broadening torch._dynamo inlining.

Benchmark Results:

Before: PYTHONPATH=/data/users/jansel/torchvision:$PYTHONPATH python benchmarks/dynamo/torchbench.py --ci --accuracy --timing --explain --print-compilation-time --disable-cudagraphs --inductor --inference --devices cuda --only vision_maskrcnn --log-graph-breaks --output /home/jansel/local/pytorch-issue-fixer/state/154259/vision_maskrcnn_before.csv produced 37 graphs, 1753 ops, 31 graph breaks, 4 unique graph breaks, and 112.507s compilation latency.

After: the same command writing vision_maskrcnn_after.csv produced 33 graphs, 1753 ops, 27 graph breaks, 3 unique graph breaks, and 107.582s compilation latency. The is_compile_supported skipped-function graph breaks disappeared; remaining local graph breaks are other existing classes.

Test Plan:

python test/dynamo/test_trace_rules.py -k test_is_compile_supported_constant

python test/dynamo/test_trace_rules.py

PYTHONPATH=/data/users/jansel/torchvision:$PYTHONPATH python benchmarks/dynamo/torchbench.py --ci --accuracy --timing --explain --print-compilation-time --disable-cudagraphs --inductor --inference --devices cuda --only vision_maskrcnn --log-graph-breaks --output /home/jansel/local/pytorch-issue-fixer/state/154259/vision_maskrcnn_after.csv

git diff --check

lintrunner -a

Fixes #154259

Generated by my agent

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says CI failed because inductor TorchBench saw the intended vision_maskrcnn graph-break improvement, reporting graph_breaks=15 while the expected baseline still had 18. I updated the inductor_torchbench expected baseline to 15 and resubmitted the ghstack PR.

**Comment by jansel** [comment]
My agent says CI failed on the inductor TorchBench training baseline because vision_maskrcnn improved to graph_breaks=33 while the expected CSV still had 38. I updated the training baseline to 33 and resubmitted. The two docs jobs built successfully and then failed at the preview upload step because the runner image had no aws CLI, so I treated those as CI environment failures rather than PR code failures.

File: test/dynamo/test_trace_rules.py [hunk 1]
```diff
@@ -416,6 +416,19 @@ def fn(x):
             res = opt_fn(x)
             self.assertEqual(ref, res)
 
+    def test_is_compile_supported_constant(self):
```
Comment (bobrenjc93) [inline test/dynamo/test_trace_rules.py]: The test only exercises `device.type == "cpu"`, where `is_compile_supported` short-circuits to `is_dynamo_supported()` and never reaches the `has_triton()` branch. Consider also (a) asserting the taken branch matches `is_compile_supported(x.device.type)` computed eagerly rather than just comparing `fn(x)` to `opt_fn(x)` (which is tautological if both take the same path), and (b) covering a `torch.device` object argument in addition to the `str`, since the signature is `DeviceLikeType` and the constant-folding path now requires `as_python_constant()` to succeed for whatever is passed.

File: torch/_dynamo/trace_rules.py [hunk 2]
```diff
@@ -189,6 +189,7 @@
     "torch.fx._symbolic_trace.is_fx_symbolic_tracing": TorchInGraphFunctionVariable,
     "torch._dynamo.external_utils.is_compiling": TorchInGraphFunctionVariable,
     "torch._dynamo.utils._disable_side_effect_safety_checks_for_current_subtracer": UserFunctionVariable,
+    "torch._dynamo.utils.is_compile_supported": UserFunctionVariable,
```
Comment (bobrenjc93) [inline torch/_dynamo/trace_rules.py:196]: Because the helper is now `_dynamo_marked_constant`, every argument must be convertible via `as_python_constant()`. If a caller passes a non-constant device (rather than the `str` torchvision passes), Dynamo now graph-breaks with `gb_type="assume_constant_result argument conversion failed"` and hints the user to "Remove torch._dynamo.assume_constant_result from this function" — advice they can't act on, since they never added the decorator. Not a regression (it graph-broke before too), but the message is misleading; consider a short comment here noting the constraint.

**Review by bobrenjc93 — APPROVED** [review]

File: test/dynamo/test_trace_rules.py (hunk 1 above)
Comment (jansel) [reply test/dynamo/test_trace_rules.py]: My agent says the following:

> I expanded the regression test to cover both a `"cuda"` string and `torch.device("cuda")`, exercising the `has_triton()` path without requiring a CUDA tensor. The expected branch is selected from an eager `is_compile_supported(device)` result rather than by calling the same wrapper eagerly.

File: torch/_dynamo/trace_rules.py (hunk 2 above)
Comment (jansel) [reply torch/_dynamo/trace_rules.py:196]: My agent says the following:

> I added a trace-rule comment documenting that tracing this marked-constant helper requires a device argument Dynamo can convert to a Python constant.

**Comment by jansel** [comment]
My agent says the following:

> I rebased onto current main, preserving its newer expected-result columns while retaining this PR's graph-break improvements (15 for inference and 33 for training). I also addressed both review comments by expanding the regression test and documenting the constant-argument constraint. The focused and full trace-rule tests, baseline checks, and `lintrunner -a` pass.
